lib/src/validator/strict_dependencies.dart

```diff
@@ -2,49 +2,76 @@
-    for (final directive in compilationUnit.directives) {
-      if (directive is UriBasedDirective) {
-        usedPackages
-            .add(Uri.parse(directive.uri.stringValue).pathSegments.first);
+class StrictDependenciesValidator extends Validator {
+  static bool _isDartFile(String file) => path.extension(file) == '.dart';
+
+  StrictDependenciesValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  /// Returns all pub packages imported or exported within [files].
+  List<_DependencyUse> _findPackages(Iterable<String> files) {
+    var allUses = <_DependencyUse>[];
+    for (var file in files) {
+      var dependencies = new List<_DependencyUse>();
+      try {
+        var contents = readTextFile(file);
+        var directives = parseImportsAndExports(contents, name: file);
+        for (var directive in directives) {
+          var usage = new _DependencyUse(directive, file, contents);
+          if (usage.isPubPackage) {
+            dependencies.add(usage);
+          }
+        }
+        allUses.addAll(dependencies);
+      } on AnalyzerErrorGroup catch (e) {
+        warnings.add('Failed parsing "$file": $e');
+      }
+    }
+    return allUses;
+  }
+
+  Future validate() async {
+    var declared = new Set<String>()
+      ..addAll(entrypoint.root.dependencies.map((d) => d.name))
+      ..addAll(entrypoint.root.devDependencies.map((d) => d.name))
+      ..add(entrypoint.root.name);
+    var allUsed = _findPackages(entrypoint.root.listFiles().where(_isDartFile));
+    for (var usage in allUsed) {
+      if (!declared.contains(usage.package)) {
+        warnings.add(usage.toErrorMessage());
       }
     }
-    packageNames[file] = usedPackages.toSet();
   }
-  return packageNames;
 }
 
-class StrictDependenciesValidator extends Validator {
-  StrictDependenciesValidator(Entrypoint entrypoint) : super(entrypoint);
+class _DependencyUse {
+  final String _contents;
+  final UriBasedDirective _directive;
+  final String _file;
+  final Uri _parsedUri;
 
-  @override
-  Future validate() {
-    return new Future.sync(() async {
-      var declared = new Set<String>()
-        ..addAll(entrypoint.root.dependencies.map((d) => d.name))
-        ..addAll(entrypoint.root.devDependencies.map((d) => d.name));
-      var allUsed = _findUsedPackages(entrypoint.root.listFiles());
-      allUsed.forEach((file, packageNames) {
-        for (var package in packageNames) {
-          if (!declared.contains(package)) {
-            warnings.add(
-                'Referenced "$package" in $file, but no dependency declared');
-          }
-        }
-      });
-    });
+  _DependencyUse(UriBasedDirective directive, this._file, this._contents)
+      : _parsedUri = Uri.parse(directive.uri.stringValue),
+        _directive = directive;
+
+  bool get isPubPackage => _parsedUri.scheme == 'package';
+
+  String get package => _parsedUri.pathSegments.first;
+
+  String toErrorMessage() {
+    return new SourceFile(_contents, url: _file)
```

Comment:
I'm guessing that you're creating the `SourceFile` here rather than sharing one between all the directives for a given file so that you avoid parsing the line endings in the common case of a file that has only valid directives? If so, mention that in a comment.